src/server/pfs/server/driver.go

```diff
@@ -806,188 +700,116 @@ func (d *driver) repoSize(ctx context.Context, repo *pfs.Repo) (int64, error) {
-		return branchInfo, nil
+	if len(branches) == 0 {
+		return nil
 	}
-
-	// subvBIMap = ( ⋃{b.subvenance | b ∈ branches} ) ∪ branches
-	// (after branches has been pruned of no-op members, which don't actually need propagation)
-	subvBIMap := map[string]*pfs.BranchInfo{}
-	for _, branch := range branches {
-		branchInfo, err := getBranchInfo(branch)
-		if err != nil {
-			return err
-		}
-
-		// We need to create new commits or aliases if any of this branch and its
-		// provenances disagree on their commit set.
-		ids := []string{branchInfo.Head.ID}
-		for _, provBranch := range branchInfo.Provenance {
-			provInfo, err := getBranchInfo(provBranch)
-			if err != nil {
-				return err
-			}
-			ids = append(ids, provInfo.Head.ID)
-		}
-		if allSameString(ids) && ids[0] != txnCtx.CommitSetID {
-			// this branch hasn't changed during the transaction, and is still consistent with its provenance,
-			// so we don't need to include it or its subvenance
-			continue
+	var propagatedBranches []*pfs.BranchInfo
+	seen := make(map[string]*pfs.BranchInfo, 0)
+	bi := &pfs.BranchInfo{}
+	for _, b := range branches {
+		if err := d.branches.ReadWrite(txnCtx.SqlTx).Get(b, bi); err != nil {
+			return errors.EnsureStack(err)
 		}
-
-		subvBIMap[pfsdb.BranchKey(branch)] = branchInfo
-		for _, subvBranch := range branchInfo.Subvenance {
-			subvInfo, err := getBranchInfo(subvBranch)
-			if err != nil {
-				return err
+		for _, sb := range bi.Subvenance {
+			if _, ok := seen[pfsdb.BranchKey(sb)]; !ok {
+				sbi := &pfs.BranchInfo{}
+				if err := d.branches.ReadWrite(txnCtx.SqlTx).Get(sb, sbi); err != nil {
+					return errors.Wrapf(err, "get subvenant branch %q", pfsdb.BranchKey(sb))
+				}
+				seen[pfsdb.BranchKey(sb)] = sbi
+				propagatedBranches = append(propagatedBranches, seen[pfsdb.BranchKey(sb)])
 			}
-			subvBIMap[pfsdb.BranchKey(subvBranch)] = subvInfo
 		}
 	}
-
-	// 'subvBIs' is the collection of downstream branches that will get a new
-	// commit. Populate subvBIs and sort it so that upstream branches are
-	// processed before their descendants (this guarantees that if branch B is
-	// provenant on branch A, we create a new commit in A before creating a new
-	// commit in B provenant on the new HEAD of A).
-	var subvBIs []*pfs.BranchInfo
-	for _, branchData := range subvBIMap {
-		subvBIs = append(subvBIs, branchData)
-	}
-	sort.Slice(subvBIs, func(i, j int) bool {
-		return len(subvBIs[i].Provenance) < len(subvBIs[j].Provenance)
+	sort.Slice(propagatedBranches, func(i, j int) bool {
+		return len(propagatedBranches[i].Provenance) < len(propagatedBranches[j].Provenance)
 	})
-
-	// Iterate through downstream branches and add new commits.
-	hasNewCommits := false
-	for _, subvBI := range subvBIs {
+	// add new commits, set their ancestry + provenance pointers, and advance branch heads
+	for _, bi := range propagatedBranches {
+		// TODO(acohen4): can we just make calls to startCommit() here?
```

Comment:
I think we should explore this. Using the same logic for creating source commits and propagation commits seems like a good idea.